In [1]:
# Cross Validation methods 
    #  Hold-out
    #  K-folds
    #  Leave-one-out
# links followed
#https://www.cs.cmu.edu/~schneide/tut5/node42.html#:~:text=Leave%2Done%2Dout%20cross%20validation,is%20made%20for%20that%20point.
#https://neptune.ai/blog/cross-validation-in-machine-learning-how-to-do-it-right

K-Fold 

k-Fold Cross-Validation: The dataset is divided into k folds (in this case, 5 folds). 
Each fold serves as a test set exactly once, and the rest of the data is used for training. 
The KFold function from scikit-learn is used to create the fold indices.

In [15]:
#imports
import pandas as pd
from sklearn.model_selection import cross_val_score, KFold
from sklearn.svm import SVC
from sklearn.datasets import load_iris
# Dataset Loading 
file_path = '/home/ex5/Desktop/AdmissionChangedDataset.csv'
df = pd.read_csv(file_path)
df.head()

SerialNo  GREScore  TOEFLScore  UniversityRating  SOP  LOR   CGPA  \
0         1       337         118                 4  4.5   4.5  9.65   
1         2       324         107                 4  4.0   4.5  8.87   
2         3       316         104                 3  3.0   3.5  8.00   
3         4       322         110                 3  3.5   2.5  8.67   
4         5       314         103                 2  2.0   3.0  8.21   

   Research  ChanceofAdmit   PassOrFail  
0         1            0.92           0  
1         1            0.76           0  
2         1            0.72           0  
3         1            0.80           0  
4         0            0.65           1

In [30]:
# Initialize a Support Vector Machine (SVM) classifier with a linear kernel
svm_classifier = SVC(kernel='linear')

# Split the dataset into features (X) and target (y)
X = df.drop(columns=["PassOrFail"])
y = df["PassOrFail"]

# Set the number of folds for k-fold cross-validation
num_folds = 5

# Initialize a KFold object for splitting the dataset into k folds
kf = KFold(n_splits=num_folds, shuffle=True, random_state=42)

# Perform k-fold cross-validation using the SVM classifier
cross_val_results = cross_val_score(svm_classifier, X, y, cv=kf)

# Print the cross-validation results
print(f'Cross-Validation Results (Accuracy): {cross_val_results}')
print(f'Mean Accuracy: {cross_val_results.mean()}')


Cross-Validation Results (Accuracy): [0.91 0.86 0.87 0.87 0.9 ]
Mean Accuracy: 0.882


Summary
The cross-validation results indicate an average accuracy of 88.2% across five folds. 
This suggests that the model performs well in predicting whether a student passes or 
fails admission based on given features.

Interpretation:
The model demonstrates strong predictive performance, with an average accuracy of 88.2%. 
This implies that it can effectively differentiate between students likely to pass or fail 
admission based on their GRE scores, TOEFL scores, university ratings, statements of purpose, 
letters of recommendation, CGPA, and research experience.

Justification:
The high mean accuracy of 88.2% suggests that the model generalizes well to unseen data, 
indicating robustness and reliability. This is crucial for admission prediction systems, 
ensuring fair evaluation of candidates based on diverse academic and personal factors. 
Such accuracy reinforces confidence in using the model for admission decision-making.

======================================================================================================================================================================================================================================================================================

Leave-One-Out Cross-Validation (LOOCV)

the Leave-One-Out Cross-Validation (LOOCV) method is implicitly utilized through the cross_val_score function when the 
number of folds (cv) is set to the length of the dataset (len(X)). LOOCV essentially creates as many folds as there 
are data points in the dataset, where each data point serves as a separate test set, and the remaining data points 
are used for training.

Leave-One-Out Cross-Validation (LOOCV) is applied by setting the cv parameter of the cross_val_score function to None. 
When cv is set to None, scikit-learn automatically performs LOOCV.

In [28]:
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LogisticRegression

# Separate features (X) and target (y)
X1 = df.drop(columns=["PassOrFail"])
X1 = df.drop(columns=["SerialNo"])
y1 = df["PassOrFail"]

# Initialize Logistic Regression model
model1 = LogisticRegression()

# Perform Leave-One-Out Cross-Validation
cross_val_results = cross_val_score(model1, X1, y1, cv=None)

# Print cross-validation results
print(f'Cross-Validation Results (Accuracy): {cross_val_results}')
print(f'Mean Accuracy: {cross_val_results.mean()}')

Cross-Validation Results (Accuracy): [1. 1. 1. 1. 1.]
Mean Accuracy: 1.0


/home/ex5/.local/lib/python3.8/site-packages/sklearn/linear_model/_logistic.py:460: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/home/ex5/.local/lib/python3.8/site-packages/sklearn/linear_model/_logistic.py:460: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_i

Interpretation:
The cross-validation results indicate perfect accuracy of 100% across all folds. 
This suggests that the model predicts whether a student passes or fails admission 
with absolute certainty based on the given features.

Justification:
The perfect mean accuracy of 100% demonstrates the model's i.e. Logistic Regression
exceptional performance in accurately classifying students as pass or fail 
based on their GRE scores, TOEFL scores, university ratings, statements of purpose, 
letters of recommendation, CGPA, and research experience. This flawless accuracy implies 
that the model effectively captures the underlying patterns in the data and makes highly 
accurate predictions without errors. Such remarkable performance underscores the robustness 
and reliability of the model, making it a valuable tool for admission decision-making.

==========================================================================================================================================

Hold-Out

In this code for the holdout method, cross-validation is not explicitly applied. 
The holdout method is a simple technique where the dataset is split into two subsets: a training set 
and a testing set. The model is trained on the training set and evaluated on the testing set.

In [29]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

# Split the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X1, y1, test_size=0.2, random_state=42)
# Initialize Logistic Regression model
model = LogisticRegression()
# Train the model on the training set
model.fit(X_train, y_train)
# Predict the target variable on the testing set
y_pred = model.predict(X_test)
# Calculate the accuracy of the model
accuracy = accuracy_score(y_test, y_pred)
# Print the accuracy
print(f'Holdout Method Results (Accuracy): {accuracy}')


Holdout Method Results (Accuracy): 1.0


/home/ex5/.local/lib/python3.8/site-packages/sklearn/linear_model/_logistic.py:460: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
